### PR TITLE
INT-3914: Add CompoundTriggerAdvice

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3914

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		return true;
+	}
+
+	@Override
+	public Message<?> afterReceive(Message<?> result, MessageSource<?> source) {
+		if (result == null) {
+			this.compoundTrigger.setOverride(this.override);
+		}
+		else {
+			this.compoundTrigger.setOverride(null);
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java]: For me it looks like `Circuit-Breaker`, like some trigger internal state, which depends on the signal.
How about this instead https://jira.spring.io/browse/INT-3679 ?

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aop/CompoundTriggerAdvice.java]: I am open to a more general solution but this is pretty trivial and easy for people to understand.

**Comment by artembilan** [comment]
Well, I can merge it as is because I don't feel well with the solution, even if it looks so simple.

From one side "Why not?" to keep it in the framework, because it is simple but may be so usable.
From other side why keep it in the framework if it is so simple? It may be just a sample to provide some food for end-users how to reach their requirements.

My concerns to the current solution:
1. The `CompoundTrigger` isn't `compound` if we don't provide `override`.
2. If it is `CompoundTrigger`, why don't require both triggers: `primary` and `secondary` as `final`?
3. Since we have two triggers there, we should just expose some state operation to switch from `primary` and `secondary` in the `CompoundTriggerAdvice` logic.

WDYT?

**Comment by garyrussell** [comment]
I see your point; I left them as volatile so you can change a cron trigger (which is immutable).

Anyway, feel free to close without merging - I provided it as a workaround for him on SO.

Just make sure that we add a note to INT-3679 to ensure it covers this use case.

**Comment by artembilan** [comment]
OK. Your `volatile so you can change a cron trigger` has convinced me :smile: .
So, merging as is. We can improve/deprecate features in the future.

**Comment by artembilan** [comment]
Merged after rebase.

**Pull request closed without merging**